meter/sma.go

```diff
@@ -184,71 +95,38 @@ func NewSMA(uri, password, iface string, serial uint32, scale float64) (api.Mete
 		}
 	}
 
-	go func() {
-		for range time.NewTicker(time.Second).C {
-			sm.updateValues()
-		}
-	}()
-
 	return decorateSMA(sm, soc), nil
 }
 
-func (sm *SMA) updateValues() {
-	sm.mux.Lock()
-	defer sm.mux.Unlock()
-
-	values, err := sm.device.GetValues()
-	if err == nil {
-		err = mergo.Merge(&sm.values, values, mergo.WithOverride)
-	}
-
-	if err == nil {
-		sm.mux.Update()
-	} else {
-		sm.log.ERROR.Println(err)
-	}
-}
-
-func (sm *SMA) hasValue() (map[sunny.ValueID]interface{}, error) {
-	elapsed := sm.mux.LockWithTimeout()
-	defer sm.mux.Unlock()
-
-	if elapsed > 0 {
-		return nil, fmt.Errorf("update timeout: %v", elapsed.Truncate(time.Second))
-	}
-
-	return sm.values, nil
-}
-
 // CurrentPower implements the api.Meter interface
 func (sm *SMA) CurrentPower() (float64, error) {
-	values, err := sm.hasValue()
-	return sm.scale * (sm.asFloat(values[sunny.ActivePowerPlus]) - sm.asFloat(values[sunny.ActivePowerMinus])), err
+	values, err := sm.device.Values()
+	return sm.scale * (sma.AsFloat(values[sunny.ActivePowerPlus]) - sma.AsFloat(values[sunny.ActivePowerMinus])), err
 }
 
 // TotalEnergy implements the api.MeterEnergy interface
 func (sm *SMA) TotalEnergy() (float64, error) {
-	values, err := sm.hasValue()
-	return sm.asFloat(values[sunny.ActiveEnergyPlus]) / 3600000, err
+	values, err := sm.device.Values()
+	return sma.AsFloat(values[sunny.ActiveEnergyPlus]) / 3600000, err
 }
 
 // Currents implements the api.MeterCurrent interface
 func (sm *SMA) Currents() (float64, float64, float64, error) {
-	values, err := sm.hasValue()
+	values, err := sm.device.Values()
 
 	measurements := []sunny.ValueID{sunny.CurrentL1, sunny.CurrentL2, sunny.CurrentL3}
```

Comment:
Magst Du das noch vereinfachen:

	var vals [3]float64
	for i, m := range []sunny.ValueID{sunny.CurrentL1, sunny.CurrentL2, sunny.CurrentL3} {
		vals[i] = sma.AsFloat(values[m])
	}

Reply:
bitte auch noch s/vals/res/, sonst sind values/vals so doppelt

Reply:
ist umgebaut und ich fand `currents` noch besser als variablen name